trunner/runners/wrappers.py

```diff
@@ -0,0 +1,339 @@
+        # TODO: investigate why send('c\r\n') does not work
+        self.proc.send('c')
+        self.proc.send('\r\n')
+
+        logging.info(f'{self.proc.before}\n')
+        self.proc.close()
+
+        if is_github_actions():
+            logging.info('::endgroup::\n')
+
+    def run(self):
+        with GdbServer() as gdbserv:
+            # Use pexpect.spawn to run a process as PTY, so it will flush on a new line
+            self.proc = pexpect.spawn(
+                self.progname,
+                self.args,
+                cwd=self.cwd,
+                encoding='ascii'
+            )
+
+            try:
+                self.read_output()
+            except (pexpect.TIMEOUT, pexpect.EOF):
+                exception = Color.colorify(f'\nGDB ERROR:', Color.BOLD)
+                exception += '\nGdb prompt not seen after executing loading script!'
+                exception = error_msg('gdb', exception, self.proc.before)
+                exception = error_msg('JlinkGdbServer', exception, gdbserv.output())
+
+                logging.info(exception)
+                sys.exit(1)
+            self.proc.close()
+
+    def close(self):
+        self.proc.close()
+        if self.proc.exitstatus != 0:
+            logging.error('gdb failed!\n')
+            raise Exception('Loading plo using gdb failed!')
+
+
+class PhoenixdError(Exception):
+    pass
+
+
+class Phoenixd(BackgroundWrapper):
+    """ Wrapper for phoenixd program"""
+
+    def __init__(
+        self,
+        target,
+        port,
+        dir='.',
+        cwd=None,
+        wait_dispatcher=True,
+        long_flashing=False
+    ):
+        if cwd is None:
+            cwd = boot_dir(target)
+        self.cwd = cwd
+        self.port = port
+        self.dir = dir
+        self.wait_dispatcher = wait_dispatcher
+        self.dispatcher_event = None
+        self.long_flashing = long_flashing
+        super().__init__()
+
+    def _reader(self):
+        """ This method is intended to be run as a separated thread. It reads output of proc
+            line by line and saves it in the output_buffer. Additionally, if wait_dispatcher is true,
+            it searches for a line stating that message dispatcher has started """
+
+        while True:
+            line = self.proc.readline()
+
+            if not line:
+                break
+            self.output_buffer += line
+
+            if self.wait_dispatcher and not self.dispatcher_event.is_set():
+                msg = f'Starting message dispatcher on [{self.port}]'
+                if msg in line:
+                    self.dispatcher_event.set()
+                    if self.long_flashing:
+                        break
+
+        # faster alternative for catching the whole output
+        if self.long_flashing:
+            last_msg = "ret=0"
+            try:
+                self.proc.expect_exact(last_msg, timeout=140)
+            # The exception will be caught in plo, we only have to provide phoenixd output
```

Comment:
Yeah, I meant that if sth was wrong with phoenixd (ret=0 not seen), the copy command wouldn't be completed successfully in plo and there would be plo exception (then output from phoenixd could be helpful)

Reply:
EDIT: I limited it to providing phoenixd output without additional warnings. I think that if sth was wrong with phoenixd, plo command would fail and we could see what was wrong in the raw phoenixd output.